### PR TITLE
feat(common): a generic long-running operation

### PR DESCRIPTION
**Pull request by coryan**
Long-running operations [aip/151] are used for API methods that take a
significant amount of time to complete (think minutes, maybe an hour).
The gRPC API returns a "promise" object, represented by the
`google::longrunning::Operation` proto, and the application (or client
library) should periodically poll this object until it is "done".

In the C++ client libraries we represent these long-running operations
by a member function that returns `future<StatusOr<ReturnType>>`. This
PR introduces a generic function to implement these member functions. It
first starts the operation using an asynchronous retry loop, and then
starts an asynchronous loop to poll the operation until it completes.

The promise can complete with an error, which is represented by a
`google::cloud::Status` object, or with success and some `ReturnType`
value.  The application may also configure the "polling policy", which
may stop the polling even though the operation has not completed.

A future PR will change the generated libraries to use this function.
With some work we could change the Spanner library to use it too,
without the need to launch detached threads to poll the results. With
substantial work we could change the Bigtable library to use it too.

[aip/151]: https://google.aip.dev/151

Part of the work for #5934 and for #4038

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6804)

### CODE REVIEW

File: google/cloud/internal/async_polling_loop_test.cc [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+  EXPECT_CALL(*mock, AsyncGetOperation).Times(0);
+  auto policy = absl::make_unique<MockPollingPolicy>();
+  EXPECT_CALL(*policy, clone()).Times(0);
+  EXPECT_CALL(*policy, OnFailure).Times(0);
+  EXPECT_CALL(*policy, WaitPeriod)
+      .WillRepeatedly(Return(std::chrono::milliseconds(1)));
+  auto actual = AsyncPollingLoop(cq, starting_op, MakePoll(mock),
+                                 std::move(policy), "test-function")
+                    .get();
+  ASSERT_THAT(actual,
```
Comment (devbww) [inline google/cloud/internal/async_polling_loop_test.cc:144]: `EXPECT_THAT`?

File: google/cloud/internal/async_polling_loop_test.cc [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  EXPECT_CALL(*policy, OnFailure)
+      .WillOnce(Return(true))
+      .WillOnce(Return(true))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*policy, WaitPeriod)
+      .WillRepeatedly(Return(std::chrono::milliseconds(1)));
+  auto actual = AsyncPollingLoop(cq, starting_op, MakePoll(mock),
+                                 std::move(policy), "test-function")
+                    .get();
+  ASSERT_THAT(actual, StatusIs(Not(Eq(StatusCode::kOk)),
```
Comment (devbww) [inline google/cloud/internal/async_polling_loop_test.cc:232]: `EXPECT_THAT`?

File: google/cloud/internal/async_polling_loop_test.cc [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  EXPECT_CALL(*policy, OnFailure)
+      .WillOnce(Return(true))
+      .WillOnce(Return(true))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*policy, WaitPeriod)
+      .WillRepeatedly(Return(std::chrono::milliseconds(1)));
+  auto actual = AsyncPollingLoop(cq, starting_op, MakePoll(mock),
+                                 std::move(policy), "test-function")
+                    .get();
+  ASSERT_THAT(actual,
```
Comment (devbww) [inline google/cloud/internal/async_polling_loop_test.cc:273]: `EXPECT_THAT`?

File: google/cloud/internal/async_long_running.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+
+/// Extracts the value (or error) from a completed long-running operation
+Status ExtractOperationResultImpl(
+    StatusOr<google::longrunning::Operation> op,
+    google::protobuf::Message& result,
+    absl::FunctionRef<bool(google::protobuf::Any const&)> validate_any,
+    std::string const& location);
```
Comment (devbww) [inline google/cloud/internal/async_long_running.h:42]: `#include <string>`

File: google/cloud/internal/async_long_running.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ * Asynchronously starts and polls a long-running operation.
+ *
+ * Long-running operations [aip/151] are used for API methods that take a
+ * significant amount of time to complete (think minutes, maybe an hour). The
+ * gRPC API returns a "promise" object, represented by the
+ * `google::longrunning::Operation` proto, and the application (or client
+ * library) should periodically poll this object until it is "done".
+ *
+ * In the C++ client libraries we represent these long-running operations by
+ * a member function that returns `future<StatusOr<ReturnType>>` this function
```
Comment (devbww) [inline google/cloud/internal/async_long_running.h:72]: s/ this/. This/

File: google/cloud/internal/async_long_running.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+ *     google::longrunning::GetOperationRequest const& request) = 0;
+ * };
+ * @endcode
+ *
+ * The corresponding `*ConnectionImpl` class would look as follows:
+ *
+ * @code
+ * class BarConnectionImpl : public BarConnection {
+ *  public:
+ *   // Using C++14 for exposition purposes, the implementation supports C++11
```
Comment (devbww) [inline google/cloud/internal/async_long_running.h:104]: s/, the/. The/
s/$/./

File: google/cloud/internal/async_long_running_test.cc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  EXPECT_THAT(*actual, IsProtoEqual(expected));
+}
+
+TEST(AsyncLongRunningTest, ExtractValueDoneWithError) {
+  google::longrunning::Operation op;
+  op.set_done(true);
+  op.mutable_error()->set_code(grpc::StatusCode::PERMISSION_DENIED);
+  op.mutable_error()->set_message("uh-oh");
+  auto const actual = ExtractLongRunningResult<Instance>(op, "test-function");
+  EXPECT_THAT(actual, StatusIs(StatusCode::kPermissionDenied));
```
Comment (devbww) [inline google/cloud/internal/async_long_running_test.cc:89]: Look for "uh-oh" too?

Comment (coryan) [reply google/cloud/internal/async_long_running_test.cc]: Done.

File: google/cloud/internal/async_long_running.h (hunk 6 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.h]: done.

File: google/cloud/internal/async_long_running.h (hunk 5 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.h]: Done.

File: google/cloud/internal/async_long_running.h (hunk 4 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.h:43]: Done.

File: google/cloud/internal/async_polling_loop_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/async_polling_loop_test.cc:273]: Done

File: google/cloud/internal/async_polling_loop.h [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * This function runs an asynchronous loop to poll the long-running operation.
+ * It periodically invokes the @p poll function to query the status of the
+ * operation. If the operation has completed (even if completed with an error)
+ * the loop stops. If the operation has not completed, or polling fails (say
+ * because the network has a problem), the function sets an asynchronous timer
+ * as configured by the polling policy and tries again later. The polling policy
+ * can stop the loop too.
+ *
+ * The function returns a `future<>` that is satisfied when the loop stops.When
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h]: Fixed the wording.

File: google/cloud/internal/async_polling_loop.h [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+ * the loop stops it satisfies the `future<StatusOr<>>` returned
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h]: Changed the wording. PTAL

File: google/cloud/internal/async_polling_loop.h [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+ * The promise can complete with an error, which is represented by a
+ * `google::cloud::Status` object, or with success and some `ReturnType` value.
+ * The application may also configure the "polling policy", which may stop the
+ * polling even though the operation has not completed.
+ *
+ * Usually the initial request to "start" the long-running operation also
+ * requires a retry loop, which is handled by this function too.
+ *
+ * Typically, library developers would use the function via
+ * `AsyncLongRunningOperation`. But as a stand-alone function it can be used in
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h]: Fixed.

File: google/cloud/internal/async_polling_loop.h [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+ * this context, first assume there is a `*Stub` class as follows:
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h]: Fixed.

File: google/cloud/internal/async_polling_loop.h [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+ *     google::longrunning::GetOperationRequest const& request) = 0;
+ * };
+ * @endcode
+ *
+ * As part of implementing a long-running operation one would do something like:
+ *
+ * @code
+ * class BarConnectionImpl : public BarConnection {
+ *  public:
+ *   // Using C++14 for exposition purposes, the implementation supports C++11
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h]: Fixed.

File: google/cloud/internal/async_polling_loop.h [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+ *    std::shared_ptr<BarStub> stub_;
+ * };
+ * @endcode
+ *
+ * [aip/151]: https://google.aip.dev/151
+ */
+future<StatusOr<google::longrunning::Operation>> AsyncPollingLoop(
+    google::cloud::CompletionQueue cq, google::longrunning::Operation op,
+    AsyncPollLongRunningOperation poll,
+    std::unique_ptr<PollingPolicy> polling_policy, std::string location);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.h:118]: Fixed.

File: google/cloud/internal/async_polling_loop.cc [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                       std::unique_ptr<PollingPolicy> polling_policy,
+                       std::string location)
+      : cq_(std::move(cq)),
+        op_(std::move(op)),
+        poll_(std::move(poll)),
+        polling_policy_(std::move(polling_policy)),
+        location_(std::move(location)) {}
+
+  future<StatusOr<google::longrunning::Operation>> Start() {
+    auto w = WeakFromThis();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.cc]: Yes, removed.

File: google/cloud/internal/async_polling_loop.cc [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    if (op_.done()) {
+      promise_.set_value(std::move(op_));
+      return promise_.get_future();
+    }
+    Wait();
+    return promise_.get_future();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_polling_loop.cc]: Done, I am not sure I like it, but I also don't think it matters that much.

File: google/cloud/internal/async_polling_loop_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/async_polling_loop_test.cc]: Done

File: google/cloud/internal/async_polling_loop_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/async_polling_loop_test.cc]: Don

**Comment by coryan** [comment]
PTAL

File: google/cloud/internal/async_long_running.cc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+Status ExtractOperationResultImpl(
+    StatusOr<google::longrunning::Operation> op,
+    google::protobuf::Message& result,
+    absl::FunctionRef<bool(google::protobuf::Any const&)> validate_any,
+    std::string const& location) {
+  if (!op) return std::move(op).status();
+  if (op->has_error()) return MakeStatusFromRpcError(op->error());
+  if (!op->has_metadata()) {
+    return Status(StatusCode::kInternal,
+                  std::string(location) +
```
Comment (devjgm) [inline google/cloud/internal/async_long_running.cc:32]: Is the explicit construction of a new `std::string` here necessary? Maybe `location` used to be a `char const*`? (One more place a few lines down too)

File: google/cloud/internal/async_long_running.h [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+ * @endcode
+ *
+ * The corresponding `*ConnectionImpl` class would look as follows:
+ *
+ * @code
+ * class BarConnectionImpl : public BarConnection {
+ *  public:
+ *   // Using C++14 for exposition purposes. The implementation supports C++11.
+ *   future<StatusOr<FooResponse>> Foo(FooRequest const& request) override {
+ *     return google::cloud::internal::AsyncLongRunningLoop(
```
Comment (devjgm) [inline google/cloud/internal/async_long_running.h:107]: Is `AsyncLongRunningLoop` the right name, or should this be `AsyncLongRunningOperation` ?

File: google/cloud/internal/async_long_running.h [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+ *  private:
+ *    google::cloud::CompletionQueue cq_;
+ *    std::shared_ptr<BarStub> stub_;
+ * };
+ * @endcode
+ *
+ * [aip/151]: https://google.aip.dev/151
+ */
+template <typename ReturnType, typename RequestType, typename StartFunctor,
+          typename RetryPolicyType>
```
Comment (devjgm) [inline google/cloud/internal/async_long_running.h:131]: Why is `RetryPolicyType` a template parameter, but `BackoffPolicy` is just a concrete type and not templated? Maybe I just forget why this is like this?

**Review by devjgm — APPROVED** [review]
A couple comments/questions, but I don't think any of mine were blockers.

File: google/cloud/internal/async_long_running.h (hunk 18 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.h:131]: There is no common base class for the Bigtable retry policies:

https://github.com/googleapis/google-cloud-cpp/blob/0687743d1881cb458c0b11a043a7d72401c07f56/google/cloud/bigtable/rpc_retry_policy.h#L81

and all the other retry policies:

https://github.com/googleapis/google-cloud-cpp/blob/0687743d1881cb458c0b11a043a7d72401c07f56/google/cloud/internal/retry_policy.h#L33

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/async_long_running.cc (hunk 16 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.cc]: Fixed, thanks.

File: google/cloud/internal/async_long_running.h (hunk 17 above)
Comment (coryan) [reply google/cloud/internal/async_long_running.h]: Fixed, also fixed the file names to match.

**Pull request merged**